src/controllers/updateTokenIcon.ts

```diff
@@ -0,0 +1,128 @@
+
+interface SignerInterface {
+  id: string;
+}
+
+interface VerifiedContract {
+  signer: SignerInterface;
+  iconUrl: String;
+  id: String;
+}
+
+interface SignedMessageRequest {
+  signMsg: JSON;
+  file: File;
+  signature: string;
+}
+
+const uploadIconRepository = config.network === 'mainnet' 
+  ? sequelize.getRepository(uploadIconMainnet) 
+  : sequelize.getRepository(uploadIconTestnet);
+
+const generateSHA256Hash = (data) => {
+  const buffer = new Uint8Array(data)
+  return crypto.subtle.digest('SHA-256', buffer).then((hashBuffer) => {
+    const hashArray = Array.from(new Uint8Array(hashBuffer))
+    const hashHex = hashArray
+      .map((b) => b.toString(16).padStart(2, '0'))
+      .join('')
+    return hashHex
+  })
+}
+
+export const uploadTokenIcon = async (
+  req: AppRequest<SignedMessageRequest>,
+  res: Response,
+) => {
+  //contract address of the token whose icon has to be updated
+  const contractAddress = toChecksumAddress(req.body.signMsg['contractAddress']);
+
+  //fetching contract details
+  const contract = await verifiedContractRepository.findByPk(contractAddress);
+  if(contract === null){
+    res.status(404);
+  }
+
+  // # FILE - CHECK 1
+  //file hash from the frontend to verify if the file sent is the same file or not
+  const fileHash = req.body.signMsg['fileHash'];
+
+  //checking if the file sent is the one which was hashed earlier
+  const file = req.body.file;
+
+  // BLOCK REPETITIVE REQUESTS
+  const uploadIconBackup = await uploadIconRepository.findByPk(contractAddress);
+  const verifiedBackup = contract;
+  try {
+    // Check if the fileHash is already present in the localdb or not
+    if (verifiedBackup) {
+      if (verifiedBackup.contractData.fileHash === fileHash) {
+        return res.status(400).json({ error: "This image is already token icon." });
+      } 
+    }
+    if(uploadIconBackup && uploadIconBackup.pending){
+      res.status(403).json({error: "Same file is being uploaded"})
+    }
+  } catch (err: any) {
+    console.error(err);
+  }
+
+  //calculating the hash for verification
+  const calculatedHash = generateSHA256Hash(file);
+  //validating the hash of file
+  if (calculatedHash != fileHash) {
+    return res.status(400).json({ error: "Invalid hash , different file uploaded" });
+  }
+
+  // # FILE - CHECK 2
+  //size of file should be less than 500kb
+  if (file.size > 500000) {
+    res.status(413).json({ error: "File size too large" });
+  }
+
+  //file is same - now checking if the icon is already uploaded or not
+  // to make it fast - checking it in localdb
+  if(verifiedBackup?.contractData.iconUrl){
+    res.status(403).json({error: "icon url already present"});
+  } 
+
+  // address of user who is trying to upload icon
+  const signerAddress = ethers.utils.verifyMessage(JSON.stringify(req.body.signMsg), ethers.utils.toUtf8Bytes(req.body.signature));
+
+
+  if (signerAddress === contract?.contractData.signer.id) {
```

Comment:
To check if `signerAddress` can update the icon we have to:

1) Check if the contract has `iconUri` field: Using Ethers, we call the `iconUri()` function of the contract (use a try-catch as the function might not exist) to check if the value exists on chain. If it exists, we do not allow to update the value.

2) Check if `signerAddress` is the owner of the contract: Using Ethers, we call the `owner()` function of the contract (use a try-catch as the function might not exist) and if it has value, get the associated evmAddress of `signerAddress` and check if it matches the address returned by `owner()`. If they do not match, we do not allow to update the value.

3) Check if `signerAddress` is the deployer of the contract: `if (signerAddress === contract.contract.signer.id)`.